### PR TITLE
Add apply_async to schedule pending tasks

### LINKED ISSUE #13512: Calling `task_a.submit(task_b.submit())` should fail with an informative error with autonomous tasks

### PR DESCRIPTION
**Pull request by abrookins**
This PR does a few things related to autonomous tasks:
1. Refactors TaskServer to use the new engine, removing the notion of task runners from `TaskServer` and autonomous tasks in general.
2. Switches the method to schedule an autonomous task from `submit()` to `apply_async()`.
3. Fixes handling of the user's retry function, which we did not honor in all cases.
4. Fixes #13512 by changing autonomous task parameter resolution to avoid descending into objects to look for `PrefectFuture` and `State` objects, since we are not tracking dependencies at this time.

**NOTE:** We should discuss separately whether it makes sense to support the equivalent of `apply_async(wait_for=[task.apply_async(1)])` or the semantically similar `apply_async(wait_for=[task.s(1)])` which in Celery orders task execution and can deliver the return result of a dependent task to the "waiting" task. Our current intention is that uses adopt flows to get features like these.

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] This pull request includes a label categorizing the change e.g. `maintenance`, `fix`, `feature`, `enhancement`, `docs`.
  

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed.

For new functions or classes in the Python SDK:

- [ ] This pull request includes helpful docstrings.
- [ ] If a new Python file was added, this pull request contains a stub page in the Python SDK docs and an entry in `mkdocs.yml` navigation.

### CODE REVIEW

File: src/prefect/engine.py [hunk 1]
```diff
@@ -1004,7 +1004,7 @@ async def orchestrate_task_run(
         # Resolve futures in parameters into data
         resolved_parameters = await resolve_inputs(parameters)
         # Resolve futures in any non-data dependencies to ensure they are ready
-        await resolve_inputs({"wait_for": wait_for}, return_data=False)
+        print(await resolve_inputs({"wait_for": wait_for}, return_data=False))
```
Comment (chrisguidry) [inline src/prefect/engine.py:1007]: stray print

**Review by chrisguidry — APPROVED** [review]
Looks...familiar :D

File: src/prefect/states.py [hunk 2]
```diff
@@ -198,7 +198,10 @@ async def exception_to_failed_state(
     #       excluded from messages for now
     message = existing_message + format_exception(exc)
 
-    return Failed(data=data, message=message, **kwargs)
+    state = Failed(data=data, message=message, **kwargs)
+    state.state_details.retriable = False
```
Comment (cicdw) [inline src/prefect/states.py:202]: nice

File: src/prefect/task_server.py [hunk 3]
```diff
@@ -225,16 +216,22 @@ async def _submit_scheduled_task_run(self, task_run: TaskRun):
             validated_state=state,
         )
 
-        self._runs_task_group.start_soon(
-            partial(
-                submit_autonomous_task_run_to_engine,
+        if task.isasync:
```
Comment (cicdw) [inline src/prefect/task_server.py:219]: this works well, but as an FYI we also have a top-level `run_task` function that does this switching on your behalf if you'd prefer - up to you

Comment (abrookins) [reply src/prefect/task_server.py:219]: Oh nice!

File: src/prefect/tasks.py [hunk 4]
```diff
@@ -1059,6 +1031,116 @@ def map(
...
+            create_task_run,
+        )
+
+        # Convert the call args/kwargs to a parameter dict
+        parameters = get_call_parameters(self.fn, args, kwargs)
+        flow_run_context = FlowRunContext.get()
+
+        # Create a pending task run
+        create_fn = create_task_run if flow_run_context else create_autonomous_task_run
+        create_task_run_call = create_call(create_fn, task=self, parameters=parameters)
```
Comment (cicdw) [inline src/prefect/tasks.py:1138]: is using the current engine primitives and utilities necessary for some reason?

**Review by cicdw — APPROVED** [review]
This LGTM, but at some point i'd love to take it further and not use the `engine.py` utilities at all if possible

File: src/prefect/tasks.py (hunk 4 above)
Comment (abrookins) [reply src/prefect/tasks.py:1138]: I'm going to see how much I can remove before merging.

**Comment by abrookins** [comment]
Eh, between benchmarks and test flakes, this is never going to pass. I'm giving it the old heave-ho.

**Pull request merged**